src/commands/deploy.ts

```diff
@@ -3,28 +3,87 @@
 
-    const folderPath: string = uri ? uri.fsPath : await workspaceUtil.selectWorkspaceFolder(ui, localize('azFunc.selectZipDeployFolder', 'Select the folder to zip and deploy'));
+    let folderPath: string = uri ? uri.fsPath : await workspaceUtil.selectWorkspaceFolder(ui, localize('azFunc.selectZipDeployFolder', 'Select the folder to zip and deploy'));
 
     if (!node) {
         node = <FunctionAppNode>(await explorer.showNodePicker(FunctionAppNode.contextValue));
     }
 
     const client: WebSiteManagementClient = getWebSiteClient(node);
+    const siteWrapper: SiteWrapper = node.siteWrapper;
+    const languageType: string = await getProjectType(folderPath);
+    if (languageType === TemplateLanguage.Java) {
+        folderPath = await getJavaFolderPath(outputChannel, folderPath, ui);
+        await updateJavaFunctionAppSettings(outputChannel, client, siteWrapper);
+    }
+
+    await siteWrapper.deployZip(folderPath, client, outputChannel);
+}
+
+async function getJavaFolderPath(outputChannel: vscode.OutputChannel, basePath: string, ui: IUserInterface): Promise<string> {
+    outputChannel.show();
+    await cpUtils.executeCommand(outputChannel, basePath, 'mvn', 'clean', 'package', '-B');
+    const targetFolder: string = path.join(basePath, 'target', 'azure-functions');
+    if (!fse.pathExistsSync(targetFolder)) {
+        throw new NoPackagedJavaFunctionError();
+    }
+    const packagedFolders: string[] = fse.readdirSync(targetFolder);
+    if (packagedFolders.length === 0) {
+        throw new NoPackagedJavaFunctionError();
+    } else if (packagedFolders.length === 1) {
+        return path.join(targetFolder, packagedFolders[0]);
+    } else {
+        return path.join(targetFolder, await promptForPackagedFolder(ui, packagedFolders));
+    }
+}
+
+async function promptForPackagedFolder(ui: IUserInterface, folders: string[]): Promise<string> {
+    const picks: Pick[] = folders.reduce(
+        (ret: Pick[], item: string): Pick[] => {
+            return ret.concat(new Pick(item));
+        },
+        []
+    );
 
-    await node.siteWrapper.deployZip(folderPath, client, outputChannel);
+    const placeHolder: string = localize('azFunc.PackagedFolderPlaceholder', 'Select packaged folder you want to deploy');
+    return (await ui.showQuickPick(picks, placeHolder, false)).label;
+}
+
+async function updateJavaFunctionAppSettings(outputChannel: vscode.OutputChannel, client: WebSiteManagementClient, siteWrapper: SiteWrapper): Promise<void> {
+    const appSettings: StringDictionary = await client.webApps.listApplicationSettings(siteWrapper.resourceGroup, siteWrapper.appName);
+    // tslint:disable-next-line:no-string-literal
+    if (appSettings['properties']['FUNCTIONS_EXTENSION_VERSION'] !== 'beta') {
+        outputChannel.appendLine(localize('azFunc.updateJavaFunctionRuntime', 'Updating FUNCTIONS_EXTENSION_VERSION to beta to support Java runtime...'));
```

Comment:
You should prompt the user before changing their function app to a beta runtime version